Improves the documentation for Azure SPs (#199)
meshcloud/meshcloud-docs: docs/meshstack.azure.index.md

5. List your Blueprint Service Principal ID via

    ```text

>     GET https://graph.microsoft.com/beta/&lt;AAD_TENANT_ID&gt;/servicePrincipals?$filter=appId eq 'f71766dc-90d9-4b7d-bd9d-4499c4331c3f'

Reviewer: d: should we still describe `#### Finding the Azure Blueprints App Id` altnerative via Graph API in our public docs? Ie what are cases in which PS `Get-AzureRmADServicePrincipal` does not work but graph API does? 

Maybe we should separate this out into a `Guide: Azure Blueprints App Id` page and link it from here so that the high-level platform integration instructions are still followable

Get-AzADServicePrincipal | Where-Object {$_.Displayname -eq "<NAME_OF_THE_SERVICE_ACCOUNT>"}
```

## Secure Azure Function Setup

In order to allow Azure Function to get called only by the replicator service principal quite a complex setup process is required which is described below.

> TODO

Reviewer 2: Should I quickly add this part?

- `Blueprint Operator` - required for assigning and managing Blueprint assignments to the Subscriptions
- `Management Group Contributor` - required on the target management group for moving Subscriptions

In order to prevent that the replicator from assigning itself more permissions than originally anticipated, it is recommended to add the following policy rule:

```json
{
    "properties": {
        "mode": "all",
        "displayName": "meshcloud Privilege Escalation Prevention",
        "description": "Prevent meshcloud SP from assigning itself new roles.",
        "policyRule": {
          "if": {
            "allOf": [
              {
                "equals": "Microsoft.Authorization/roleAssignments",
                "field": "type"
              },
              {
                "allOf": [
                  {
                    "field": "Microsoft.Authorization/roleAssignments/roleDefinitionId",

>                     "equals": "/subscriptions/<SUBSCRIPTION_ID>/providers/Microsoft.Authorization/roleDefinitions/<AZURE_SERVICE_PRINCIPAL_ROLE_ID>"

Reviewer: d: This may be wrong, i.e. AZURE_SERVICE_PRINCIPAL_ROLE_ID is the **expected** role assignment. Shouldn't this policy audit any **unexpected** role assignments? should this be `notEquals` instead of `equals` (check with azure policy language, I'm not 100% familiar with it)

                  },
                  {
                    "field": "Microsoft.Authorization/roleAssignments/principalId",
                    "equals": "<AZURE_SERVICE_PRINCIPAL_ID>"
                  },
                ]
              }
            ]
          },
          "then": {
            "effect": "audit"
          }
      }
    }
}
```

> If you intend to call Azure functions via a Landing Zone configuration, then you need to grant the replicator principal also permissions to the role you have created for the [Azure secured function call](#secure-azure-function-setup).

Reviewer: f: `secure-azure-function-setup` this link does not exist

> #### Azure Function Access Setup

Reviewer: d: should we move this into it's own page?

    ```json
    {
      "allowedMemberTypes": [
        "Application"
      ],
      "description": "Allows an SPP to get a token to a restricted application",
      "displayName": "SPP-Access",
      "id": "<RANDOM_UUID>",
      "isEnabled": true,
      "lang": null,
      "origin": "Application",
      "value": "Access"
    }
    ```

    ![App Role Manifest](assets/azure_function/app-role-manifest.png)

4. In your Service Principals API permissions screen, add the newly created Application Role. Don't forget to grant admin consent again afterwards.

    ![Assign the Application Role to SP](assets/azure_function/sp-token.png)

After these steps, you should be able to fetch a token scoped to this Application Role (via the Service Principal secrets). The scope field inside the JWT token should match the application's ID. With this token you can call the Azure Function. This is the same functionality that also the meshReplicator is using.

![Fetch Token](assets/azure_function/fetch-token.png)

![Decoded Token](assets/azure_function/decoded-token.png)

> #### meshFed Identity Lookup

Reviewer: d: this belongs into Identity Lookup configuration page, not here where we set up meshPlatform

3. argument: meshProject [ID (numeric)](./meshstack.configuration.md#identifiers)
4. argument: tenantNumber (numeric), a running number specific to each platform which can optionally be enabled
5. argument: role name suffix ([configurable via Landing Zone](./meshstack.azure.landing-zones.md#meshrole-to-platform-role-mapping))

## Permission Model
<!-- markdownlint-disable MD001 -->

meshstack in total needs up to three service principals, one for replicating meshProjects into the Azure platform, one if AAD should be used as an user lookup when users are added to a customer or project inside the meshfed panel and one for [gathering metering data](./meshstack.azure.metering.md) for the billing subsystem.

#### meshReplicator Project Provisioning

- `Directory.Read.All` - this permission is required to search the directory for existing users, groups and service principals
- `Group.ReadWrite.All` - this permissions is required to create new groups
- `User.Invite.All` - this permission is required if you want to enable B2B User Invitations
- `User Access Administrator` - to add and remove user group access to the Subscriptions

> - `Rights to call Azure Functions` - see the [end of this section](#azure-function-access-setup) for how to setup this

Reviewer 2: ```suggestion
- (optional)`Permissions to call Azure Functions` - see the [end of this section](#azure-function-access-setup) for how to setup this
```

> Created Subscriptions will have the Service Principal of the replicator registered as owner at first. As soon as all needed maintainance steps are performed (like renaming the subscription, moving it to the final management group etc.), the replicator removes itself as an owner.

Reviewer 2: ```suggestion
Created Subscriptions will have the Service Principal of the replicator registered as an owner at first. As soon as all needed maintenance steps are performed (e.g. renaming the subscription, moving it into the final management group), the replicator removes itself as an owner.
```

> All permissions left are therefore granted only via the management group hierarchy. The meshstack software does **not** need access related to actual workload inside this subscriptions, however to perform certain maintenance tasks the following permissions/roles must be granted to the replicator principal through the management group hierarchy:

Reviewer 2: ```suggestion
All permissions left are therefore granted only via the management group hierarchy. The meshstack software does **not** need access related to actual workload inside these subscriptions. However in order to perform certain maintenance tasks, the following permissions/roles must be granted to the replicator principal:
```

- `Blueprint Operator` - required for assigning and managing Blueprint assignments to the Subscriptions
- `Management Group Contributor` - required on the target management group for moving Subscriptions

> In order to prevent that the replicator is able to assign himself more permissions than originally anticipated its recommended to add a policy rule like this:

Reviewer 2: ```suggestion
In order to prevent that the replicator from assigning itself more permissions than originally anticipated, it is recommended to add the following policy rule:
```

```json
{
    "properties": {
        "mode": "all",
        "displayName": "meshcloud Privilege Escalation Prevention",
        "description": "Prevent meshcloud SP from assigning itself new roles.",
        "policyRule": {
          "if": {
            "allOf": [
              {
                "equals": "Microsoft.Authorization/roleAssignments",
                "field": "type"
              },
              {
                "allOf": [
                  {
                    "field": "Microsoft.Authorization/roleAssignments/roleDefinitionId",
                    "equals": "/subscriptions/<SUBSCRIPTION_ID>/providers/Microsoft.Authorization/roleDefinitions/<AZURE_SERVICE_PRINCIPAL_ROLE_ID>"
                  },
                  {
                    "field": "Microsoft.Authorization/roleAssignments/principalId",
                    "equals": "<AZURE_SERVICE_PRINCIPAL_ID>"
                  },
                ]
              }
            ]
          },
          "then": {
            "effect": "audit"
          }
      }
    }
}
```

> If you intend to call Azure functions via a Landing Zone configuration then you need to grant the replicator principal also permissions to the role you have created for the Azure secured function call, described [here](#secure-azure-function-setup).

Reviewer 2: ```suggestion
If you intend to call Azure functions via a Landing Zone configuration, then you need to grant the replicator principal also permissions to the role you have created for the [Azure secured function call](#secure-azure-function-setup).
```

#### Azure Function Access Setup

In order to make an Azure Function only accessible via the replicators Service Principal follow these steps:

> 1. Create a system assigned identity for your function.

Reviewer 2: This step is not required for the function access. This only allows your function to have permissions inside Azure (like starting VMs, Connecting LogWorkspaces and such)

    ![System assigned identity](assets/azure_function/system-assigned-identity.png)

2. Lock down your function to only allow assigned users.

    ![Assigned users only](assets/azure_function/assigned-users.png)

3. Create a custom [Application Role](https://docs.microsoft.com/en-us/azure/architecture/multitenant-identity/app-roles). Its only possible to assign real users and unfortunatly no Service Principals directly to the function so this additional steps are required. Edit the Application Roles manifest like in this JSON:

    ```json
    {
      "allowedMemberTypes": [
        "Application"
      ],
      "description": "Allows an SPP to get a token to a restricted application",
      "displayName": "SPP-Access",
      "id": "<RANDOM_UUID>",
      "isEnabled": true,
      "lang": null,
      "origin": "Application",
      "value": "Access"
    }
    ```

    ![App Role Manifest](assets/azure_function/app-role-manifest.png)

4. In your Service Principals API permissions screen, add the newly created Application Role. Don't forget to grant admin consent again afterwards.

    ![Assign the Application Role to SP](assets/azure_function/sp-token.png)

> When these steps are successful, you should be able to fetch a token scoped to this Application Role (via the Service Principal secrets), and this token should contain a scop to the app. With this token you can call the Azure Function.

Reviewer 2: ```suggestion
After these steps, you should be able to fetch a token scoped to this Application Role (via the Service Principal secrets). The scope field inside the JWT token should match the application's ID. With this token you can call the Azure Function. This is the same functionality that also the meshReplicator is using.
```

![Fetch Token](assets/azure_function/fetch-token.png)

![Decoded Token](assets/azure_function/decoded-token.png)

#### meshFed Identity Lookup

> This principal is only required if the AAD is actually be used as source for users information when assigning users to customers or projects inside the meshPanel. If this should be used then the following permissions are required on this principal:

Reviewer 2: ```suggestion
This principal is only required if the AAD is actually be used as source of user information when assigning users to customers or projects inside the meshPanel. In order to use this functionality, the principal requires the following permissions:
```
